examples/fluids/src/setupstg_shur14.c

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
+// reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+/// @file
+/// Implementation of the Synthetic Turbulence Generation (STG) algorithm
+/// presented in Shur et al. 2014
+
+#include <stdlib.h>
+#include <math.h>
+#include "../qfunctions/stg_shur14.h"
+
+#ifndef M_PI
+#define M_PI    3.14159265358979323846
+#endif
+
+#ifndef NOMINMAX
+
+#ifndef max
+#define max(a,b)            (((a) > (b)) ? (a) : (b))
+#endif
+
+#ifndef min
+#define min(a,b)            (((a) < (b)) ? (a) : (b))
+#endif
+
+#endif  /* NOMINMAX */
+
+/*
+ * @brief Perform Cholesky decomposition on array of symmetric 3x3 matrices
+ *
+ * This assumes the input matrices are in order [11,22,33,12,13,23]. This
+ * format is also used for the output.
+ *
+ * @param[in] nprofs Number of matrices in Rij
+ * @param[in] Rij Array of the symmetric matrices [6,nprofs]
+ * @param[out] Cij Array of the Cholesky Decomposition matrices, [6,nprofs]
+ */
+static inline void CalcCholeskyDecomp(int nprofs,
+                                      const CeedScalar Rij[6][nprofs], CeedScalar Cij[6][nprofs]) {
+
+  CeedPragmaSIMD
+  for(int i=0; i<nprofs; i++) {
+    Cij[0][i] = sqrt(Rij[0][i]);
+    Cij[3][i] = Rij[3][i] / Cij[0][i];
+    Cij[1][i] = sqrt(Rij[1][i] - pow(Cij[3][i], 2) );
+    Cij[4][i] = Rij[4][i] / Cij[0][i];
+    Cij[5][i] = (Rij[5][i] - Cij[3][i]*Cij[4][i]) / Cij[1][i];
+    Cij[2][i] = sqrt(Rij[2][i] - pow(Cij[4][i], 2) - pow(Cij[5][i], 2));
+  }
+}
```

Comment:
Not to fall into the premature optimization trap, but I have a general question about n-dimensional array layouts. Above, I have the index being looped over as the last index so `Cij[5][0]` is non-contiguous to `Cij[4][0]` despite being looped together. This is against my general intuition, which is to keep data that is used in same loop iteration contiguous.  Looking at the assembly of the above function with both array orientations (`[6][nprofs]` and `[nprofs][6]`), having the looped index be the first array index (`[nprofs][6]`) results in less total operations and move vectorized operations. 

However, I wrote the function with `[6][nprofs]` orientation as this is what QFunctions with n-dimensional array data do (just replace `nprofs` with `CEED_Q_VLA`) https://github.com/CEED/libCEED/blob/9477ba7d4a95edebf665b0c36c17bb403727d847/examples/fluids/qfunctions/eulervortex.h#L314-L319.

Is there reason that the QFunctions use (to my eye) a less easily vectorizable data structure? Is my view point too simplified? Is it a limitation of VLA?

Reply:
For our state variables, we want the quadrature point last so every operation is vectorized (uses SIMD instructions) over that dimension. Since all quadrature points are independent, that's good. For memory locality, you want data that you use together to be close. That's why (on CPU), the QFunction gets called in batches of elements that fit in (usually L1 or L2) cache. All the constant size loops (usually 3 or 5) inside the QFunction get unrolled, and the loop over quadrature points also gets unrolled (usually by 4 or 8) to use these vector instructions.

Some operations will prevent the compiler from vectorizing. We hit that with `log1p` in solids, and wrote our own log1p that's accurate enough in the relevant range. Variable length inner loops are another that will often make the compiler give up on vectorizing. We can profile and optimize this if it's an issue. The main way to do that would be to sum the series at a batch of points (e.g., 8), ensuring vectorization over that inner loop. With that, your `Cij` would only need to be accessed as scalars anyway, thus the ordering won't really matter, and interlacing the components (`[nprofs][6]`) will give somewhat better L1 cache reuse.

Reply:
If you want to do it that way, I would make a struct with `Cij` and the other terms indexed up to `nprofs`, then the other. It'll be fewer offsets (maybe just two) and thus slightly fewer integer registers needed for indexing. (Very much premature optimization, but the code might be more clear if written this way.) I would just focus on writing clear code right now -- I have a couple theories on how best to optimize it, but that'll be easier if we can test our hypothesis with working code, and also, it might not need that much attention.

Reply:
Ok, that makes sense. I'll probably just leave the structures as is for right now, though I agree it'd probably make the code a little more clear.